add generated action interface

vhdirk opened the pull request:
Let's start off with the action interface, since that one is used by application at some point.
This generated file is not at all usable yet, but let's see where we get from here.

vhdirk committed:
add generated action interface

vhdirk committed:
regenerate

sdroege commented on gio-subclass/src/action.rs:
```diff
@@ -0,0 +1,220 @@
+    iface_data: glib_ffi::gpointer
+) {
+    callback_guard!();
+    let action_iface = &mut *(iface as *mut ffi::);
```
This is missing the C interface struct type here after `ffi::`

sdroege commented on gio-subclass/src/action.rs:
```diff
@@ -0,0 +1,220 @@
+
+unsafe extern "C" fn action_get_type<T: ObjectType>(
+    type_: glib_ffi::GType
+) -> glib::Type {
```
This function is not needed for interfaces, what you basically have here is a trampoline calling a `get_type` function on the interface... but there is none on `GAction`

sdroege commented on gio-subclass/src/action.rs:
```diff
@@ -0,0 +1,220 @@
+    let interface_static = &*(ptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    imp.get_state_type()
```
Like signal trampolines, this needs all the conversion stuff for the return value (`from_glib_full` in this case) and also the parameters (none here). You also want to pass the `&gio::Action` via `&from_glib_borrow(ptr)` to this and all other functions

sdroege commented on gio-subclass/src/action.rs:
```diff
@@ -0,0 +1,220 @@
+    let interface_static = &*(ptr as *const T::InstanceStructType);
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    imp.change_state()
```
Here you want to pass both `ptr` and `value` to the function, which should work the same as in signal trampolines

sdroege commented on gio-subclass/src/action.rs:
```diff
@@ -0,0 +1,220 @@
+
+pub trait ActionImpl: AnyImpl + 'static {
+
+    fn activate(&self, action: &T, parameter: Option<&glib::Variant>);
```
The `&T` here should be a `&gio::Action` AFAIU

vhdirk replied:
True. I was thinking for a moment that we could make actionImpl generic over T? But i'm not sure if that has any benefits to it

sdroege replied:
That has the problem that you can't use `Box<ActionImpl>` anymore but have to provide the type parameter everywhere, or not?

vhdirk replied:
Not sure. I can't even remember why I wanted to make it generic

sdroege replied:
Let's not do that for now then :) It can always be dynamically casted if needed

sdroege commented:
Looks generally very good, the main thing missing here at this point seems to be the type conversion between C and Rust, which should be the same as in the signal trampolines

vhdirk committed:
regenerate

vhdirk committed:
regenerate

vhdirk committed:
regenerate

vhdirk commented:
Aside from missing imports and so on, the largest issue remains those trampolines. I've been looking at the signals implementation, but I can only find stuff marked as TODO :)

vhdirk committed:
regenerate

sdroege commented:
> I've been looking at the signals implementation, but I can only find stuff marked as TODO :)

There's https://github.com/gtk-rs/gir/blob/master/src/codegen/trampoline.rs, specifically the  `trampoline_call_func` function. `trampoline_call_parameters` is converting all the parameters from C to Rust, `analysis.ret.trampoline_to_glib` converts the return value back to C.

vhdirk committed:
regenerate

vhdirk commented:
Thanks. I'm not sure how correct this is, but it seems like it could work

sdroege commented on gio-subclass/src/action.rs:
```diff
@@ -0,0 +1,202 @@
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap: T = from_glib_borrow(instance);
+    imp.activate(&wrap, &from_glib_none(action), /*Unknown conversion*/parameter)
```
This could be `&from_glib_borrow(action)` (and IIRC would be for signals). Not sure why it falls apart on the `GVariant`, in the function just below it works :)

sdroege replied:
Generally this should do `&from_glib_borrow` for all shared/boxed/object types (if `transfer none`) like it would in signal trampolines. There's probably a code path that has to be enabled for your case too here

sdroege replied:
See e.g. https://github.com/sdroege/gstreamer-rs/blob/85ca3b94a34d0236314aac900d5926c81ba73d99/gstreamer/src/auto/element.rs#L535-L540

vhdirk replied:
Actually, all of the 'action' conversions here are incorrect. That's what 'wrap' is

sdroege replied:
Oh! Indeed

sdroege commented on gio-subclass/src/action.rs:
```diff
@@ -0,0 +1,202 @@
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap: T = from_glib_borrow(instance);
+    imp.get_name(&wrap, &from_glib_none(action))/*Not checked*/.to_glib_none().0
```
The return value is unsafe but that's because the C API is broken, this needs a manual implementation.

You need to return a string here that is valid for the whole lifetime of the underlying action object

sdroege replied:
@federicomenaquintero things like this are btw examples why I think we will need an override system somewhere between gnome-class and the .gir files, and why I believe the trampolines and some glue infrastructure should live in the gtk/etc bindings for gnome-class to use.

I have some more examples in GStreamer :) We should talk about this at the next hackfest in more detail

sdroege commented on gio-subclass/src/action.rs:
```diff
@@ -0,0 +1,202 @@
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap: T = from_glib_borrow(instance);
+    imp.get_parameter_type(&wrap, &from_glib_none(action))/*Not checked*/.to_glib_none().0
```
Same as for the name

sdroege commented on gio-subclass/src/action.rs:
```diff
@@ -0,0 +1,202 @@
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap: T = from_glib_borrow(instance);
+    imp.get_state(&wrap, &from_glib_none(action)).to_glib_full()
```
This return value could use an optimization but not important right now
```rust
let ret = imp.get_state(...);
let ptr = ret.to_glib_none().0;
mem::forget(ret);
ptr
```

sdroege commented on gio-subclass/src/action.rs:
```diff
@@ -0,0 +1,202 @@
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap: T = from_glib_borrow(instance);
+    imp.get_state_type(&wrap, &from_glib_none(action))/*Not checked*/.to_glib_none().0
```
Same problem as for name

vhdirk committed:
regenerate

vhdirk committed:
regenerate

vhdirk committed:
regenerate

vhdirk committed:
regenerate

sdroege commented:
> So either I fix that manually, or we come up with a plan to pull implementations from a separate file.

Generally I was thinking separate files should work fine, e.g. marking a vfunc as ignore in the toml file and then having one file per vfunc that defines the 3 parts of it: 1) the trait function signature, 2) the chaining up part, 3) the trampoline. And then during code generation, those parts are spliced into the output.

vhdirk commented:
Do you have any proposition regarding the syntax of that file? I'd like to make it as simple as possible to know what content to splice where. Ideally, the gir should not know about what's in that file, only which block to insert where.

sdroege commented:
> Do you have any proposition regarding the syntax of that file? I'd like to make it as simple as possible to know what content to splice where. Ideally, the gir should not know about what's in that file, only which block to insert where.

Yes I agree, ideally we wouldn't need to parse it at all. So maybe just separate files with the actual content that is inserted in the relevant places? `manual/action_vfuncname_trampoline.custom`, etc?

vhdirk committed:
regenerate (with content spliced from external file)

vhdirk commented:
Something like this? (see gir for implementatin details)

sdroege commented:
Yes, exactly that :)

vhdirk committed:
regenerate

vhdirk committed:
regenerate

sdroege commented:
```diff
diff --git a/gio-subclass/src/action.rs b/gio-subclass/src/action.rs
index 7542553..9a6a927 100644
--- a/gio-subclass/src/action.rs
+++ b/gio-subclass/src/action.rs
@@ -78,9 +78,9 @@ unsafe extern "C" fn action_activate<T: ObjectType>
     let param = if parameter.is_null(){
         None
     }else{
-        Some(&from_glib_none(parameter))
+        Some(from_glib_none(parameter))
     };
-    imp.activate(&wrap, param)
+    imp.activate(&wrap, param.as_ref())
 }
 
 unsafe extern "C" fn action_change_state<T: ObjectType>
@@ -206,7 +206,7 @@ unsafe extern "C" fn action_init<T: ObjectType>(
     iface_data: glib_ffi::gpointer
 ) {
     callback_guard!();
-    let action_iface = &mut *(iface as *mut gio_ffi::GAction);
+    let action_iface = &mut *(iface as *mut gio_ffi::GActionInterface);
     let iface_type = (*(iface as *const gobject_ffi::GTypeInterface)).g_type;
     let type_ = (*(iface as *const gobject_ffi::GTypeInterface)).g_instance_type;
     let klass = &mut *(gobject_ffi::g_type_class_ref(type_) as *mut ClassStruct<T>);
```

This is required to make it compile. Wrong type is used for the interface, and some lifetime issues with a variable

vhdirk committed:
regenerate

vhdirk commented:
Thanks!

vhdirk commented:
So while this thing now compiles, I have no idea how to use it. I'd guess we need ```SimpleAction``` or so in order to test this?

sdroege commented:
No, `SimpleAction` already *implements* the `GAction` interface. We'd need a new `GObject` and then can have it implement the interface. I'll write a compiling-but-not-working example for that later :)

sdroege committed:
Implement minimal test for the GAction interface

Criticals during initialization as we don't implement all properties
needed by the interface, and is also non-functional because of
unimplemented virtual methods.

sdroege commented:
You can find an example that compiles here. Let me know if you have any questions about the code, and also note the function I've added to `action.rs`

vhdirk committed:
regenerate

vhdirk committed:
regenerate

sdroege commented:
Were you able to write a subclass of `GAction` that is actually doing something meaningful?

vhdirk commented:
No. Not yet. I don't know what to do with that one either :)

sdroege commented:
Ah, I thought you started with `GAction` because that's something you actually needed :) I guess look at the implementation of `GSimpleAction` in gio then

vhdirk commented:
I started with action since application implements action group and actionmap. So I figured I'd work my way up. Meanwhile I've almost finished generating application itself.

vhdirk commented:
I'm working on recreating a minimal version of ```GSimpleAction``` and an adaptation of its tests. I'll probably do the same for ```GActionMap``` and ```GActionGroup``` later.

vhdirk committed:
rebase on master

vhdirk committed:
wip: proper test for gaction

vhdirk commented:
I've started writing a better test, based on the C ```SimpleAction```. However, I think the generated code is not really correct. Seems to me that the parameter ```action: &gio::Action``` is actually unneeded?
Furthermore I'm missing some glib stuff to pass a ```VariantType``` as a ```Value```

sdroege commented:
> However, I think the generated code is not really correct.

Which parts seem wrong?

> Seems to me that the parameter action: &gio::Action is actually unneeded?

Without this parameter, you have no access to the actual `GObject` from any of your virtual functions. You can't call any e.g. `GAction` API on anything then.

You might not need it currently, but in other cases it is needed.

> Furthermore I'm missing some glib stuff to pass a VariantType as a Value

For storing a `GVariant` in a `GValue` and getting it back from there again?

vhdirk commented:
> Without this parameter, you have no access to the actual GObject from any of your virtual functions. You can't call any e.g. GAction API on anything then.
> 
> You might not need it currently, but in other cases it is needed.

In the test, I want to call ```action.get_enabled()```. But even though our simpleaction implements the action interface, it does not seem to have the method required? Do I need to downcast manually?

> For storing a GVariant in a GValue and getting it back from there again?
Not a ```GVariant```. A ```GVariantType```

sdroege commented on gio-subclass/tests/simple_action.rs:
```diff
@@ -0,0 +1,264 @@
+
+glib_wrapper! {
+    pub struct SimpleAction(Object<imp::SimpleAction>):
+        [Object => InstanceStruct<Object>];
```
> In the test, I want to call action.get_enabled(). But even though our simpleaction implements the action interface, it does not seem to have the method required? Do I need to downcast manually?

You're missing the `gio::Action => gio_ffi::GAction` part here

vhdirk replied:
Ah yes. Thanks

vhdirk committed:
add gaction to glib_wrapper list

sdroege commented:
> You mean I have to implement those and submit a PR?

Yes, those trait impls are needed for being able to put something into `GValue`s. If you look at the other impls, they're really simple so don't worry :) It's just some translation glue between Rust and GLib land

vhdirk committed:
wip: beef up gaction test

This still panics when using the getter on a Boxed VariantType though

vhdirk committed:
fix simpleaction unit test

vhdirk commented:
I've extended the unit test for a bit. It does not test everything (not by a long shot), but it's a start.
It does need https://github.com/gtk-rs/glib/pull/340 though

vhdirk commented:
What's annoying here is that the unsafe bits prevent debugging of panic's. For instance, when a method in ```impl ActionImpl ``` is ```unimplemented!````, the last meaningful information you can find in the backtrace is from the trampolines callback guard.
Isn't there anything we can do to improve this?
Can we 'repanic' the backtrace from ```thread:: panicked``` in the callback guard?

vhdirk committed:
extend simpleaction unit test some more

vhdirk committed:
wip: simpleaction test

sdroege commented:
If you set `RUST_BACKTRACE=1` then you should still get a backtrace of the original panic. It might however be useful to add something with https://crates.io/crates/backtrace to the callback guard.

vhdirk committed:
wip: simpleaction test

sdroege commented:
In this context also see https://github.com/gtk-rs/gir/pull/614

vhdirk commented:
enabling the backtrace is usually my first thing to try. But that doesn't give any useful clues either. I'll just use gdb until gtk-rs/gir#614 is merged.

Can you look into my last commit? it panics in 2 places, that I feel are more related to glib's c stuff than anything else.

vhdirk committed:
invert did_run assertion

vhdirk committed:
invert get_property(enabled) assertion

sdroege commented:
> enabling the backtrace is usually my first thing to try. But that doesn't give any useful clues either. I'll just use gdb until gtk-rs/gir#614 is merged.

Weird, never had such problems in the past. Maybe some behaviour changed



> Can you look into my last commit? it panics in 2 places, that I feel are more related to glib's c stuff than anything else.

Sure, will take a look some time this weekend :)

vhdirk commented:
The backtraces were _really_ hard to read. It was just that I had 2 typos in assertions (I had them negated again). that's fixed now and the tests pass.

sdroege commented:
Ah great. Anything else I should take a look at already? :)

sdroege commented on gobject-subclass/src/properties.rs:
```diff
@@ -16,6 +16,8 @@ pub enum PropertyMutability {
     Writable,
     ReadWrite,
+    Construct,
+    ConstructOnly
```
This should be in a separate commit, and has to be solved a bit differently. `Construct` properties can be read and/or write, `ConstructOnly` properties can be readable (after construction) or they can be neither. In C these are all flags that can be combined in arbitrary ways.

The rules are here https://gitlab.gnome.org/GNOME/glib/blob/a65ee63c711111d61deb336da9ed9a13ce0126e8/gobject/gobject.c#L539-554, apart from that it can be combined in any possible way

vhdirk replied:
I figure we need the bitflags crate here?

sdroege replied:
Or we express the possible combinations with one or two enums. Not all combinations are valid, so bitflags are not so nice either.

sdroege replied:
Oh I see the problem... hm!

sdroege replied:
Can this part be backed out for now until the PR that implements it properly?

vhdirk replied:
Sure, it's only for the test anyway

sdroege commented on gobject-subclass/src/properties.rs:
```diff
@@ -102,6 +106,14 @@ pub enum Property<'a> {
         fn() -> glib::Type,
         PropertyMutability,
     ),
+    Variant(
```
We need to add support for `Variant`s in `Value`s in GLib for this first, you only added `VariantType` so far :)

vhdirk replied:
I think that's already there though, because of ```glib_wrapper```.

sdroege replied:
Indeed, but it's doing it wrong unfortunately. The macro considers it a boxed type but it isn't anymore: `Note: GLib 2.24 did include a boxed type with this name. It was replaced with this fundamental type in 2.26.`

sdroege replied:
https://github.com/gtk-rs/glib/issues/344

sdroege commented on gio-subclass/src/auto/action.rs:
```diff
@@ -0,0 +1,238 @@
+use gobject_subclass::object::*;
+
+
+pub trait ActionImpl: AnyImpl + 'static {
```
Also needs to be `ObjectImpl`. The docs say that `GAction requires GObject` (this should also be machine-readable from the gir file)

vhdirk replied:
True. For interfaces I did not expect there to be requirements as such

vhdirk replied:
This does pose a problem though. ```ObjectImpl``` is generic over T, but ```ActionImpl``` isn't?

sdroege replied:
Hmm, could we make it generic over `T` or does that break everything then?

vhdirk replied:
It'll probably break a lot, but I don't see any other option? Though I'm not sure what traits ```T``` should be bound to.

sdroege replied:
Would have to be bound to `ObjectImpl`. Let's worry about that in a separate PR

sdroege commented on gio-subclass/src/auto/action.rs:
```diff
@@ -0,0 +1,238 @@
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap = from_glib_borrow(gptr);
+    match imp.get_name(&wrap){ Some(t)  => t/*Not checked*/.to_glib_none().0, None => ptr::null()}
```
This is unfortunately unsafe (and `get_parameter_type()` and `get_state_type()`). The pointers returned need to be valid after returning (by using `to_glib_none` they stop being valid after the scope of the function is gone).

A possible solution for this would be doing it like the protocol stuff in `GstURIHandler`, see https://github.com/sdroege/gst-plugin-rs/blob/754373ba6c89cc2bb9fe5779b156dcf0e374a465/gst-plugin/src/uri_handler.rs#L32 (and wherever it is used). Here these "static" values are part of the `Static` part of the interface and it ensures that the underlying pointers are always valid.

However for your case you need them to be valid per instance, not per type (correct?). So a solution to that would be to require them to be given during instance initialization (by having some public function that has to be called then), and to store them externally in the instance so that the implementation can't modify them. E.g. via `g_object_get_qdata` and the `g_object_set_qdata_full`. That way you have control over the lifetime of the pointers instead of the implementation.

vhdirk replied:
Pfew. I'm not sure how I could get that to look clean though...

sdroege replied:
Or you could do it all in the trampoline for the vfunc. You check if some qdata already exists, if so you just use that. Otherwise (i.g. the first time) you call the impl for the return value and then store it in the qdata.

vhdirk replied:
That seems like a better approach to me. Seems silly to store a return value somewhere in user-impls. As far as rust is concerned, the callee should no longer care about the value.

sdroege replied:
This could in theory also be autogenerated from the code generator, but better just do it manually for the cases for now and let's see if we find a weird case.

sdroege commented on gio-subclass/tests/simple_action.rs:
```diff
@@ -0,0 +1,391 @@
+                None => {
+                    //FIXME: how can we propagate the optional nature of the arg?
+                    //       the value created here seems dangling for glib, and panics
+                    action.emit("activate", &[&glib::Variant::from(false)]).unwrap();
```
You could pass `None` I guess

vhdirk replied:
Not really. The emit function is rather strict on its argument. It has to be something that implements to_value. And it has to be a ref.

sdroege replied:
How about `&None::<glib::Variant>`?

vhdirk replied:
Wow. I can try that. I'm not even sure I understand what that means :)

sdroege replied:
If you just write `&None` then the compiler tries to infer of what type `T` that `Option<T>` should be, but here it has no information at all what it could be. So you need to provide that type yourself. It's like with `let stuff = some_iter.collect::<Vec<_>>()` where you tell the compiler that you'd like to have a `Vec` and not something else that can be collected into

sdroege commented on gio-subclass/tests/simple_action.rs:
```diff
@@ -0,0 +1,391 @@
+        // FIXME: This needs some simplification...
+        //
+        // Get us Send/Sync constraints
+        let f: Box<Fn(&Self, Option<glib::Variant>) + Send + Sync + 'static> = unsafe {
```
Fancy! `GVariant` is be definition `Send` and `Sync` (it's immutable after all). That should be fixed in GLib

vhdirk replied:
This may not be needed. Might be the result of some hasty copy paste work

vhdirk replied:
It is needed like it is there, but not because of ```GVariant```. Rather because of ```self``` which is not ```Send``` + ```Sync```. Does it need to be?

sdroege replied:
Can you nonetheless submit a PR for `GVariant`? It will probably appear again sooner or later for real :)

For `Self`, it's only needed because we can't statically know if the signal involves other threads. In this case it is safe what you're doing here.

vhdirk commented:
I wonder how much of this test is needed in order to verify the correctness of the generated interface. I was not planning on duplicating ```GSimpleAction``` in full here.

vhdirk committed:
wip: ownership of pointers

vhdirk committed:
fix argument for activate signal

vhdirk committed:
regenerate

vhdirk commented:
@sdroege Can you take a look at my latest additions? I've moved handling of ```Option``` return types into the gir, and I've also started working on storing pointers in the quark data where ownership is not transferred.
Destroying those pointers is not yet implemented. The easiest thing would be to call ```from_glib_full``` and just let the new object go out of scope, but that's a bit costly I presume. What I'd like to do here is split off the ```free```ing part of ```from_glib_full``` into a new trait ```FreeGLibPtr```, so that that one can be reused. The other option would be to call the correct ```free```ing function each time, but that would bring more complexity into gir.

vhdirk committed:
is this a good way to implement those free functions?

sdroege commented on gio-subclass/src/auto/action.rs:
```diff
@@ -123,7 +125,7 @@ unsafe extern "C" fn action_get_name<T: ObjectType>
                 glib_ffi::g_quark_from_string("rs_get_name".to_glib_none().0),
                 ret as *mut c_void,
-                None //TODO: how do we free the data
+                Some(<String as FreeGlibPtr>::free)
```
I would simply define a local function here that does the freeing instead of defining a new trait

sdroege replied:
Oh I see the problem... hm!

vhdirk committed:
second try for destroy function

vhdirk commented on gio-subclass/src/auto/action.rs:
```diff
@@ -118,14 +118,19 @@ unsafe extern "C" fn action_get_name<T: ObjectType>
                 glib_ffi::g_quark_from_string("rs_get_name".to_glib_none().0),
                 ret as *mut c_void,
-                Some(<String as FreeGlibPtr>::free)
+                Some(destroy)
```
Would you prefer to see it like this?
I guess that works too, but I find that that would end up as a lot of duplicate code. Does the gir know what ```free``` function to call? Like ```get_type```? Because if not, I don't think we can do much different than ```FreeGlibPtr```

sdroege replied:
Apart from `from_glib_full()` and having things go out of scope, I don't see any other solution either. And for strings that involves another copy until I fix that, for other types it should be fine.

So maybe we can go with `from_glib_full()` for now and assume the string problem is solved soon. It's on my list for the next hackfest

vhdirk replied:
So I'll go with a nested destroy function that calls ```from_glib_full```.

sdroege replied:
That's what I would do for now, and if there's a performance problem we can always implement that manually until the string problem is solved

sdroege commented on gio-subclass/src/auto/action.rs:
```diff
@@ -0,0 +1,274 @@
+    let imp = instance.get_impl();
+    let imp = (*(*interface_static).imp_static).get_impl(imp);
+    let wrap = from_glib_borrow(gptr);
+    imp.activate(&wrap, (if parameter.is_null() { None } else { Some(from_glib_none(parameter)) }).as_ref())
```
You want `Some(&from_glib_borrow(parameter))` here, and also below `from_glib_borrow()` instead of `_none()`

vhdirk replied:
Haha, I made the current implementation as per your previous review :).
I can do ```from_glib_borrow``` here. What's the difference?

sdroege replied:
Potentially fewer copies :)

sdroege commented on gio-subclass/src/auto/action.rs:
```diff
@@ -0,0 +1,274 @@
+        Some(t) => {
+            let ret = t/*Not checked*/.to_glib_none().0;
+            gobject_ffi::g_object_set_qdata_full(gptr as *mut gobject_ffi::GObject,
+                glib_ffi::g_quark_from_string("rs_get_name".to_glib_none().0),
```
Should probably also contain the interface/object name in the quark string

sdroege commented on gio-subclass/src/auto/action.rs:
```diff
@@ -0,0 +1,274 @@
+    match imp.get_name(&wrap)
+    {
+        Some(t) => {
+            let ret = t/*Not checked*/.to_glib_none().0;
```
This needs to be `to_glib_full()` so that the return value stays valid forever until you free it

sdroege commented on gio-subclass/tests/simple_action.rs:
```diff
@@ -0,0 +1,386 @@
+            "The type of the state kept by the action",
+            glib::VariantType::static_type,
+            PropertyMutability::ReadWrite //|
+            // PropertyMutability::ConstructOnly |
```
This needs some fixing

vhdirk replied:
I know. I'm going to leave that for now and address that in a separate PR.

sdroege replied:
I'm surprised the test succeeds nonetheless, but yes let's do that separately :)

vhdirk committed:
regenerate

vhdirk committed:
regenerate

vhdirk committed:
regenerate

vhdirk commented:
I think I've adressed all issues within the scope of this PR?

vhdirk committed:
Fix lib after removal of 'free' module

sdroege commented:
> I think I've adressed all issues within the scope of this PR?

Thanks, will go over it later today again :)

sdroege commented on gio-subclass/src/auto/action.rs:
```diff
@@ -0,0 +1,266 @@
+    imp_static: *const ActionImplStatic<T>
+}
+
+// FIXME: Boilerplate
```
This can go away I guess

vhdirk replied:
Just for this interface, or for any?

sdroege replied:
It's generated code so I guess can generally go away

vhdirk replied:
Are you talking about the ```FIXME``` part, or ```imp_static```. I was thinking the latter, but the FIXME is a wee bit silly indeed :)

sdroege replied:
The FIXME

vhdirk committed:
remove new GParamSpec properties for now

vhdirk committed:
regenerate

vhdirk committed:
regenerate

vhdirk commented:
Not sure why there are merge conflicts all of a sudden. I did rewrite the git index because I sometimes forget to use the correct email address in my git config.

sdroege commented:
Try rebasing directly on master and then force-push. Then it should be fine.

vhdirk committed:
regenerate

vhdirk commented:
I've regenerated one last time. It now includes some ```let rs_ret``` statements. That's because I've been tailoring the gir to support writing back output parameters to pointers. I'll get a PR using that specifically up soon.

sdroege commented:
Ok, go for it then :)

vhdirk commented:
Ready for merge?

sdroege commented:
Yes, but please don't forget to look at the property flag stuff afterwards :)

vhdirk commented:
I'll do that soon. I'll need it anyway.
